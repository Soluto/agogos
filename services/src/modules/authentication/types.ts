import { ClientOptions } from 'jwks-rsa';

export interface IssuerConfig {
  authority: string;
  jwksConfig?: Partial<ClientOptions>;
  audience?: string | string[];
  authenticatedPaths: string[];
  description?: string;
}

export interface AuthenticationConfig {
  jwt?: Record<string, IssuerConfig>;
  anonymous?: {
    publicPaths: string[];
    rejectAuthorizationHeader?: boolean;
  };
}
