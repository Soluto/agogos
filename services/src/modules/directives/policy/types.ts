import {PolicyArgsObject, PolicyAttachments} from '../../resource-repository/types';

export type Policy = {
    namespace: string;
    name: string;
    args?: PolicyArgsObject;
};

// args here contain the final values after param injection
export type PolicyExecutionContext = {
    namespace: string;
    name: string;
    policyAttachments: PolicyAttachments;
    args?: PolicyArgsObject;
    query?: QueryResults;
};

export type QueryResults = {
    [name: string]: any;
};

export type PolicyExecutionResult = {
    done: boolean;
    allow?: boolean;
    query?: {
        type: string;
        code: string;
    };
};

export type GraphQLArguments = {
    [name: string]: any;
};
