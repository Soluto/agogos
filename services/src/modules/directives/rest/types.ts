export interface KeyValue {
  key: string;
  value: string;
}

export interface RestParams {
  url: string;
  method?: string;
  body?: string;
  bodyArg?: string;
  query?: KeyValue[];
  headers?: KeyValue[];
  timeoutMs?: number;
}
