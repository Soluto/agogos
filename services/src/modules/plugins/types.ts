import { BaseSchema } from '../base-schema';
import { ResourceGroup } from '../resource-repository';

export type ValueOrPromise<T> = T | Promise<T>;

export interface StitchPlugin {
  name: string;

  addArgumentInjectionGlobals?(): ValueOrPromise<Record<string, unknown>>;

  transformResourceGroup?(resourceGroup: ResourceGroup): ValueOrPromise<ResourceGroup>;

  transformBaseSchema?(baseSchema: BaseSchema): ValueOrPromise<BaseSchema>;
}
