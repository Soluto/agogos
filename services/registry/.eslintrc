{
    "env": {
        "es6": true,
        "mocha": true,
        "jest/globals": true
    },
    "extends": [
        "standard"
    ],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": 2018,
        "sourceType": "module"
    },
    "plugins": [
        "@typescript-eslint",
        "jest"
    ],
    "rules": {
        "@typescript-eslint/no-unused-vars": "error",
        "space-before-function-paren": "off",
        "space-in-parens": "off",
        "array-bracket-spacing": "off",
        "semi": "off",
        "@typescript-eslint/semi": ["error"],
        "jest/no-disabled-tests": "warn",
        "jest/no-focused-tests": "error",
        "jest/no-identical-title": "error",
        "jest/prefer-to-have-length": "warn",
        "jest/valid-expect": "error"
    },
    "overrides": [{
        "files": ["tests/**"],
        "rules": {
            "no-unused-expressions": "off"
        }
    }]
}