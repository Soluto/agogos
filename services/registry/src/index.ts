import nconf = require("nconf");
import logger, { loggingMiddleware } from "./logger";
nconf.env("__");

import * as express from "express";
import gqlSchemaRoute from "./routes/schema";
import validateRoute from "./routes/validate";
import { startGrpcServer } from "./sync/sync-grpc-server";

const PORT = process.env.PORT || 4000;

const app = express();

app.use(loggingMiddleware);

app.get("/health", (_: express.Request, res: express.Response) =>
    res.send(true),
);
app.get("/metrics", (_: express.Request, res: express.Response) =>
    res.send(true),
);

app.use("/schema", gqlSchemaRoute);
app.use("/validate", validateRoute);

app.listen({ port: PORT }, () =>
    logger.info(`ðŸš€ HTTP Server ready at http://localhost:${PORT}`),
);

startGrpcServer();
