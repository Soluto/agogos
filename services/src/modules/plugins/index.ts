export { createBasicPolicyPlugin } from './base-policy';
export { createLoggingPlugin } from './logging';
export { createMetricsPlugin } from './metrics';
export { createArgumentInjectionPlugin } from './argument-injection';
