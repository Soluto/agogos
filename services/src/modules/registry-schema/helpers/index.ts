import proceedPolicy from './policy-helper';

export { proceedPolicy };
