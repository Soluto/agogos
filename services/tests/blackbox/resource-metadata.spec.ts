import * as fs from 'fs/promises';
import { print } from 'graphql';
import { FastifyInstance } from 'fastify';
import { gql } from 'apollo-server-core';
import { createServer as createRegistry } from '../../src/registry';
import { updateSchemasMutation } from '../helpers/registry-request-builder';
import { ResourceGroup, Schema } from '../../src/modules/resource-repository';
import { PluginMetadata } from '../../src/modules/plugins/types';

const plugins: PluginMetadata[] = [
  {
    name: 'First',
    version: '0.1.0',
  },
  {
    name: 'Second',
    version: '1.2.3',
  },
];
jest.mock('../../src/modules/plugins', () => ({
  ...(jest.requireActual('../../src/modules/plugins') as Record<string, unknown>),
  getPlugins: jest.fn(() => plugins),
}));

describe('Resource metadata', () => {
  let app: FastifyInstance;

  const schema1: Schema = {
    metadata: {
      namespace: 'blackbox-tests',
      name: 'resource-metadata-1',
    },
    schema: print(gql`
      type Query {
        foo: String
      }
    `),
  };

  const schema2: Schema = {
    metadata: {
      namespace: 'blackbox-tests',
      name: 'resource-metadata-2',
    },
    schema: print(gql`
      type Query {
        bar: String
      }
    `),
  };

  const registryResourceFile = process.env.FS_REGISTRY_RESOURCE_REPOSITORY_PATH!;
  const registryResourceMetadataFile = process.env.FS_REGISTRY_RESOURCE_METADATA_PATH!;

  const gatewayResourceFile = process.env.FS_RESOURCE_REPOSITORY_PATH!;
  const gatewayResourceMetadataFile = process.env.FS_RESOURCE_METADATA_PATH!;

  beforeAll(async () => {
    app = await createRegistry();

    const resources: ResourceGroup = {
      schemas: [],
      upstreams: [],
      upstreamClientCredentials: [],
      policies: [],
      remoteSchemas: [],
    };
    await fs.writeFile(process.env.FS_REGISTRY_RESOURCE_REPOSITORY_PATH!, JSON.stringify(resources));
  });

  afterAll(async () => {
    await app.close();
    await fs.unlink(gatewayResourceFile);
    await fs.unlink(registryResourceFile);
    await fs.unlink(registryResourceMetadataFile);
    await fs.unlink(gatewayResourceMetadataFile);
  });

  test('First update', async () => {
    const responseFoo = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload: {
        query: updateSchemasMutation,
        variables: {
          schema: schema1,
        },
      },
    });
    expect(responseFoo.statusCode).toEqual(200);
    expect(responseFoo.json().data.result.success).toBeTruthy();
  });

  test('Check metadata files', async () => {
    const registryResourceMetadata = await fs.readFile(registryResourceMetadataFile, { encoding: 'utf8' });
    expect(registryResourceMetadata).toMatchSnapshot();

    const gatewayResourceMetadata = await fs.readFile(gatewayResourceMetadataFile, { encoding: 'utf8' });
    expect(gatewayResourceMetadata).toMatchSnapshot();
  });

  test('Second update', async () => {
    const responseFoo = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload: {
        query: updateSchemasMutation,
        variables: {
          schema: schema2,
        },
      },
    });
    expect(responseFoo.statusCode).toEqual(200);
    expect(responseFoo.json().data.result.success).toBeTruthy();
  });

  test('Check metadata files again', async () => {
    const registryResourceMetadata = await fs.readFile(registryResourceMetadataFile, { encoding: 'utf8' });
    expect(registryResourceMetadata).toMatchSnapshot();

    const gatewayResourceMetadata = await fs.readFile(gatewayResourceMetadataFile, { encoding: 'utf8' });
    expect(gatewayResourceMetadata).toMatchSnapshot();
  });
});
