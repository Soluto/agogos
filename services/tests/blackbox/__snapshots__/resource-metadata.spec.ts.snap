// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource metadata Check metadata files 1`] = `
"{
  \\"checksum\\": \\"83d4b8c0a958a17fc6ba922b99476413\\",
  \\"summary\\": {
    \\"schemas\\": 1,
    \\"upstreams\\": 0,
    \\"upstreamClientCredentials\\": 0,
    \\"policies\\": 0,
    \\"remoteSchemas\\": 0,
    \\"policyAttachments\\": 1
  }
}"
`;

exports[`Resource metadata Check metadata files 2`] = `
"{
  \\"checksum\\": \\"0ea9e8d688ca6f7bc46eb8cee9071a40\\",
  \\"summary\\": {
    \\"schemas\\": 1,
    \\"upstreams\\": 0,
    \\"upstreamClientCredentials\\": 0,
    \\"policies\\": 0,
    \\"remoteSchemas\\": 0,
    \\"policyAttachments\\": 1,
    \\"pluginsData\\": 1
  },
  \\"plugins\\": [
    {
      \\"name\\": \\"First\\",
      \\"version\\": \\"0.1.0\\"
    },
    {
      \\"name\\": \\"Second\\",
      \\"version\\": \\"1.2.3\\"
    }
  ]
}"
`;

exports[`Resource metadata Check metadata files again 1`] = `
"{
  \\"checksum\\": \\"0c798a1f566dc58d635e503363121e05\\",
  \\"summary\\": {
    \\"schemas\\": 2,
    \\"upstreams\\": 0,
    \\"upstreamClientCredentials\\": 0,
    \\"policies\\": 0,
    \\"remoteSchemas\\": 0,
    \\"policyAttachments\\": 1
  }
}"
`;

exports[`Resource metadata Check metadata files again 2`] = `
"{
  \\"checksum\\": \\"9ea80e577c5cb0df653380727e0e4325\\",
  \\"summary\\": {
    \\"schemas\\": 2,
    \\"upstreams\\": 0,
    \\"upstreamClientCredentials\\": 0,
    \\"policies\\": 0,
    \\"remoteSchemas\\": 0,
    \\"policyAttachments\\": 1,
    \\"pluginsData\\": 1
  },
  \\"plugins\\": [
    {
      \\"name\\": \\"First\\",
      \\"version\\": \\"0.1.0\\"
    },
    {
      \\"name\\": \\"Second\\",
      \\"version\\": \\"1.2.3\\"
    }
  ]
}"
`;
